### PR TITLE
Add connector section to the specification document

### PR DESCRIPTION
**Pull request by cescoffier**
Write the connector section in the spec.

Related to #22.

### CODE REVIEW

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 1]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
 The implementation is reponsible for enforcing the acknowledgement strategy defined by the user when the `@Acknowledgement` policy is used. 
 If the annotation is not used, the default policy must be enforced.
 
+=== Connector
+
+Reactive Messaging connects matching `@Incoming` and `@Outgoing` running inside the same application.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: needs a noun, suggest "stream elements"

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 2]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+Besides, it maps specific _channels_ to _external_ technologies such as Apache Kafka, MQTT, Web Sockets, or JMS.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "Besides" -> "As well as this" or "Additionally"

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 3]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
 The implementation is reponsible for enforcing the acknowledgement strategy defined by the user when the `@Acknowledgement` policy is used. 
 If the annotation is not used, the default policy must be enforced.
 
+=== Connector
+
+Reactive Messaging connects matching `@Incoming` and `@Outgoing` running inside the same application.
+Besides, it maps specific _channels_ to _external_ technologies such as Apache Kafka, MQTT, Web Sockets, or JMS. 
+It means that Reactive Messaging can receive messages from virtually any messaging technology and dispatch messages to any messaging technology.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "It" -> "This"

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 4]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+The component responsible for this role is called _connector_.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: Should be
"The components responsible for this role are called _connectors_."
or
"This bridging to an external messaging technology is done using a  reactive messaging _connector_."

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 5]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+Besides, it maps specific _channels_ to _external_ technologies such as Apache Kafka, MQTT, Web Sockets, or JMS. 
+It means that Reactive Messaging can receive messages from virtually any messaging technology and dispatch messages to any messaging technology.
+The component responsible for this role is called _connector_. 
+
+==== Connector concepts
+
+Each _connector_ is responsible for a specific technology. 
+A connector can:
+
+* act as a _Publisher_, meaning it retrieves or receives messages from an external messaging technology. The messages will then be sent to a method annotated with `@Incoming`
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "technology." -> "technology and publishes them to a reactive stream."

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 6]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+* act as a _Subscriber_, meaning it dispatches messages to an external messaging technology. The messages are received from a method annotated with `@Outgoing`
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "meaning it" -> "meaning it subscribes to a reactive stream and"

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 7]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+* handle both directions
+
+It's essential that connectors implement the back-pressure protocol defined by the Reactive Streams specification.
+
+A _connector_ is implemented as a CDI Bean, generally _application_ scoped implementing:
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: Does it have to implement 1 of or both of these interfaces?

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 8]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+
+* the `org.eclipse.microprofile.reactive.messaging.connector.IncomingConnectorFactory` interface to receive messages from an external source;
+* the `org.eclipse.microprofile.reactive.messaging.connector.OutgoingConnectorFactory` interface to dispatch messages to an external sink
+
+The bean is a factory called by the Reactive Messaging implementation to create `PublisherBuilder` or `SubscriberBuilder` objects. 
+These objects are then connected to methods annotated with `@Incoming` or `@Outgoing`. 
+
+Both interfaces contains a `type()` method returning a `MessagingProvider`.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "contains" -> "contain"

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 9]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+import org.eclipse.microprofile.reactive.messaging.MessagingProvider;
+
+public class Kafka implements MessagingProvider {
+
+  // Just a marker class.
+
+}
+----
+
+Once defined, the user can, in the configuration, refer to this connector using the fully qualified name of the class.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: I would like to discuss at the hangout if it is advantageous for this to have
get...ConnectionFactory methods to reduce the amount of class loading
of ConnectionFactories that might not be used.

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 10]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+==== Configuration
+
+Reactive Messaging connectors are configured using MicroProfile Config.
+The implementation processes the global configuration and determines:
+
+* which channels are defined
+* which connectors are used (using the `type`) attribute
+* the configuration for each channel
+
+The builder methods defined in the `IncomingConnectorFactory` and `OutgoingConnectorFactory` receives a `org.eclipse.microprofile.config.Config` as parameter.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "receives" -> "receive"

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 11]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+The `Config` object contains key-value pairs to configure the connector.
+The configuration is specific to the connector.
+Typically, a Kafka connector expects a _bootstrap servers_ entry as well as a _topic_ entry.
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "Typically," -> "For example,"

File: spec/src/main/asciidoc/architecture.asciidoc [hunk 12]
```diff
@@ -976,6 +974,115 @@ public Message<O> process(Message<I> msg) {
...
+ mp.messaging.incoming.my-channel.topic=my-topic
+---- 
+
+The Reactive Messaging implementation:
+
+1. Reads the configuration
+2. Identifies that a `my-channel` source needs to be managed
+3. Searches for the `type` attribute and finds `org.eclipse.reactive.sample.kafka.Kafka`
+4. Looks for a bean implementing the `IncomingConnectorFactory` interface and having a `type()` method returning the `org.eclipse.reactive.sample.kafka.Kafka` class. If the configuration had contained a `mp.messaging.outgoing.my-channel...` entry, a bean implementing the `OutgoingConnectorFactory` interface would have been searched.
+5. Creates a new `Config` object with just the `key=value` pairs:
```
Comment (hutchig) [inline spec/src/main/asciidoc/architecture.asciidoc]: "the key" -> "the relevant key"

File: spec/src/main/asciidoc/architecture.asciidoc (hunk 7 above)
Comment (cescoffier) [reply spec/src/main/asciidoc/architecture.asciidoc]: Depends on your connector. Generally, both except... when you integrate read-only mechanism (dbus) or write-only systems (machine learning).

Adding a sentence about this.

**Pull request merged**